tools/extract_challenges.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python
+
+'''Extract challenges from lessons.'''
+
+import sys
+import re
+
+
+USAGE = 'usage: extract_challenges [-o output_name] [filename...]'
+
+title_pat = re.compile(r'^>\s+.+\{.challenge\}')
+block_pat = re.compile(r'^>')
+
+def main(argv):
+    '''Main driver.'''
+
+    prog_name = argv[0]
+    filenames = argv[1:]
+    output_name = None
+    writer = sys.stdout
+
+    if filenames and (filenames[0] == '-o'):
+        if len(filenames) < 2:
+            fail(USAGE)
+        output_name = filenames[1]
+        filenames = filenames[2:]
+
+    if output_name:
+        writer = open(output_name, 'w')
+
+    if filenames:
+        for f in filenames:
+            with open(f, 'r') as reader:
+                extract(reader, sys.stdout)
+    else:
+        extract(sys.stdin, sys.stdout)
```

Comment:
That's in line with UNIX conventions, but less common in python utilities, where the default behavior tends to be printing a help message. 

I'm not sure there's a strong use case for supporting piping to stdin for this utility; help text might be preferable.